### PR TITLE
fix: unicode is handled properly in strdist

### PR DESCRIPTION
**Pull request by letFunny**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----
While doing more performance work I noticed a couple of bugs in our implementation of distance. I will fix each one of them in a different PR so that the perf work can land. I have added a TODO for the next bug. I don't want to fix both as part of the same PR as it will be much harder to reason about the code, the fix for unicode in itself is pretty easy.

This PR fixes a common bug when handling strings in Go. The for loop:
```go
for bi, br := range b {
    ...
    lst[bi+1] = cost
}
```
Is iterating over a string and using `bi` which is the byte offset to store things in `lst` instead of the current rune count. In the case where the characters are ASCII it works because each rune is 1 byte but in the case of runes that take more than 1 byte this fails (see the added test case).

### CODE REVIEW

File: internal/strdist/strdist.go [hunk 1]
```diff
@@ -87,8 +87,7 @@ func Distance(a, b string, f CostFunc, cut int64) int64 {
 		if debug {
 			debugf("... %v", lst)
 		}
-		_ = stop
-		if cut != 0 && stop {
+		if cut != 0 && len(b) > 0 && stop {
```
Comment (letFunny) [inline internal/strdist/strdist.go:90]: This is not strictly related to unicode but it made the test fail when the second string was empty so I fixed it as well.

For context, `stop` is meant to represent that the minimum edit cost is greater than the threshold. However, if the second string is empty, `stop` will not be set to false by the inner loop, which is a bug (see test case).

**Review by upils — APPROVED** [review]
Thanks! Would you mind reworking the PR description to clarify what is actually fixed? I found it a little tricky to spot.

File: internal/strdist/strdist.go (hunk 1 above)
Comment (niemeyer) [reply internal/strdist/strdist.go]: Can you put into words why this is the right fix here?  Note that this is disabling the cut logic altogether when b is empty.

File: internal/strdist/strdist_test.go [hunk 2]
```diff
@@ -58,6 +58,7 @@ var distanceTests = []distanceTest{
 	{f: strdist.GlobCost, r: 1, a: "a**f/hij", b: "abc/def/hik"},
 	{f: strdist.GlobCost, r: 2, a: "a**fg", b: "abc/def/hik"},
 	{f: strdist.GlobCost, r: 0, a: "a**f/hij/klm", b: "abc/d**m"},
+	{f: strdist.GlobCost, r: 1, a: "**a", b: ""},
```
Comment (niemeyer) [inline internal/strdist/strdist_test.go:67]: Can we please have an independent test for the empty b as well?

**Review by niemeyer — COMMENTED** [review]
Thanks for the fix! A question about the added logic.

**Review by niemeyer — CHANGES_REQUESTED** [review]
👆

File: internal/strdist/strdist_test.go (hunk 2 above)
Comment (letFunny) [reply internal/strdist/strdist_test.go:67]: Added

File: internal/strdist/strdist.go (hunk 1 above)
Comment (letFunny) [reply internal/strdist/strdist.go]: Your intuition was right this is only a partial fix. I added several more test cases that were failing previously. I also changed the logic to finish as soon as `cut` is reached. The issue was the following, `stop` is computed by iterating over `b[i]` and seeing if:
* cost of swapping it for `a[i]`
* cost of inserting `b[i]`
* cost of removing `a[i]`
* 0 if `a[i] == b[i]`

any of these was less than `cut`. The problem is that to enter the loop `b` could not be empty. The calculation above the for loop correctly accounted for this case in `lst[0]`, that is comparing `a[:j]` to `b[:0]`. The problem is that we were not using the cost here to compute `min`.

You were the one devising the algorithm so I hope I got a good understanding, let me know if we need to discuss it in more depth in person.

File: internal/strdist/strdist_test.go [hunk 3]
```diff
@@ -58,7 +59,11 @@ var distanceTests = []distanceTest{
...
 	{f: strdist.GlobCost, r: 0, a: "/*a/", b: "/a/"},
+	{f: strdist.GlobCost, r: 3, a: "abc", b: ""},
+	{f: strdist.GlobCost, r: 1, cut: 1, a: "abc", b: ""},
+	{f: strdist.GlobCost, r: 2, cut: 3, a: "ab", b: ""},
```
Comment (upils) [inline internal/strdist/strdist_test.go]: These tests are unrelated to using globs. What about testing them with `StandardCost` instead to make it obvious?

**Review by upils — APPROVED** [review]
We discussed offline some aspects of this function that this work uncovered (non-symmetric behavior,  potential optimization) but I think this PR should focus on fixing the unicode-related bug, so it is good as-is. Thanks!

File: internal/strdist/strdist_test.go (hunk 3 above)
Comment (letFunny) [reply internal/strdist/strdist_test.go]: Done, let me know what you think. I also added more tests following's @niemeyer suggestion in the sprint.

File: internal/strdist/strdist_test.go [hunk 4]
```diff
@@ -44,6 +45,11 @@ var distanceTests = []distanceTest{
 	{f: strdist.StandardCost, r: 3, a: "abcdefg", b: "axcdfgh"},
 	{f: strdist.StandardCost, r: 2, cut: 2, a: "abcdef", b: "abc"},
 	{f: strdist.StandardCost, r: 2, cut: 3, a: "abcdef", b: "abcd"},
+	{f: strdist.StandardCost, r: 3, a: "abc", b: ""},
+	{f: strdist.StandardCost, r: 1, cut: 1, a: "abc", b: ""},
+	// Not symmetric.
+	{f: strdist.StandardCost, r: 2, cut: 3, a: "ab", b: ""},
+	{f: strdist.StandardCost, r: 2, cut: 1, a: "", b: "ab"},
```
Comment (upils) [inline internal/strdist/strdist_test.go:52]: suggestion: add another comment after to make it clear the non-symmetry is only illustrated with these 2 cases?

File: internal/strdist/strdist_test.go [hunk 5]
```diff
@@ -58,7 +64,15 @@ var distanceTests = []distanceTest{
...
 	{f: strdist.GlobCost, r: 0, a: "a**f/hij/klm", b: "abc/d**m"},
+	{f: strdist.GlobCost, r: 1, a: "**a", b: ""},
 	{f: strdist.GlobCost, r: 0, a: "/*a/", b: "/a/"},
+	// Edge cases with empty strings.
+	{f: strdist.GlobCost, r: strdist.Inhibit, cut: 1, a: "/", b: ""},
+	{f: strdist.GlobCost, r: strdist.Inhibit, cut: 1, a: "", b: "/"},
+	{f: strdist.GlobCost, r: 0, cut: 1, a: "*", b: ""},
+	{f: strdist.GlobCost, r: 0, cut: 1, a: "", b: "*"},
+	{f: strdist.GlobCost, r: 0, cut: 1, a: "**", b: ""},
+	{f: strdist.GlobCost, r: 0, cut: 1, a: "", b: "**"},
```
Comment (upils) [inline internal/strdist/strdist_test.go:75]: Unrelated to the fix but still very welcome to have these additional tests, thanks!

**Review by upils — APPROVED** [review]
Thanks @letFunny. I like that the additional tests give us a clearer view of the behavior for the corner cases.

File: internal/strdist/strdist_test.go (hunk 4 above)
Comment (letFunny) [reply internal/strdist/strdist_test.go:52]: I thought about doing that but I could not think of a way of being idiomatic. The usual is `// End of symmetric tests` or something like that but there are no more sections in the code. In Chisel we use a summary but that too fells heavyweight.

File: internal/strdist/strdist.go [hunk 6]
```diff
@@ -28,20 +28,28 @@ func StandardCost(ar, br rune) Cost {
...
 }
 
+// Distance returns the edit distance between two strings. The cost per edit is
+// given by the costFunc argument.
+//
+// There is an optional cut argument that when set will finish the computation
+// as early as possible once the final cost is certain to be >= cut. There is
+// no guarantee about the exact cost returned when this is the case other than
+// being >= cut. In particular, when cut is used, the function is not symmetric
+// on a and b.
```
Comment (niemeyer) [inline internal/strdist/strdist.go:38]: It's missing information on how to disable cut. Just says it's optional, but it's not actually optional.

**Review by niemeyer — APPROVED** [review]
Thanks for the fix and the additional test.

We need to fix that detail in the comment, but I'll commit so the fix is in.

Can you please also submit this on go-algo itself?

Thanks again.

**Review by niemeyer — CHANGES_REQUESTED** [review]
Sorry, I'm changing my review and holding it as I need a commit message as well.

**Comment by niemeyer** [comment]
Nevermind, I've picked up the bottom half of the description. Just needs the comment fix and porting to go-algo.
